### PR TITLE
unstable-book: split sanitizers into testing and production ones

### PR DESCRIPTION
**Pull request by disconnect3d**
The Rust Unstable Book lists compiler sanitizers that can be used in Rust programs. However, it does not say whether it is okay to use certain sanitizers on production or not, which may suggest that its okay to do so.

I believe that none of ASAN/TSAN/MSAN/LeakSAN should be used on production.

There was an old thread on oss-security that provided more details here: https://www.openwall.com/lists/oss-security/2016/02/17/9 but one example to not use those sanitizers on production is the fact that some of them use environment variables that control things like "path to the binary that will be used to get the symbol names for stack traces". As a result, if a sanitized binary has suid, an attacker can use the specific environment variable to run their own program and escalate their privileges this way.

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
r? rust-lang/compiler

**Comment by Nilstrieb** [comment]
Thanks, adding some disambiguation is great! I think this is correct but I'm not entirely sure, I'd rather let someone who knows more about this approve it.
r? compiler

File: src/doc/unstable-book/src/compiler-flags/sanitizer.md [hunk 1]
```diff
@@ -9,20 +9,22 @@ The tracking issues for this feature are:
...
-* [LeakSanitizer](#leaksanitizer) a run-time memory leak detector.
-* [MemorySanitizer](#memorysanitizer) a detector of uninitialized reads.
-* [MemTagSanitizer](#memtagsanitizer) fast memory error detector based on
-  Armv8.5-A Memory Tagging Extension.
-* [ShadowCallStack](#shadowcallstack) provides backward-edge control flow protection.
-* [ThreadSanitizer](#threadsanitizer) a fast data race detector.
+  * [LeakSanitizer](#leaksanitizer) a run-time memory leak detector.
+  * [MemorySanitizer](#memorysanitizer) a detector of uninitialized reads.
+  * [ThreadSanitizer](#threadsanitizer) a fast data race detector.
+* Those that apart from testing, may be used on production:
```
Comment (est31) [inline src/doc/unstable-book/src/compiler-flags/sanitizer.md:19]:
```suggestion
* Those that apart from testing, may be used in production:
```

**Comment by eholk** [comment]
This change is probably fine, although I think it might be a little out of scope for the Rust Unstable Book. It seems reasonable for us to say "these are sanitizers you can use" and direct people to further documentation on those sanitizers to evaluate whether it makes sense to use it in production or just testing.

**Comment by disconnect3d** [comment]
Maybe we should have then a sentence or two at the beginning explaining what a sanitizer is, that would explain something along the lines that they are compiler plugins which enable additional instrumentation of the code that may be useful for testing/security testing and sometimes for production, depending on the sanitizer.

**Comment by apiraino** [comment]
I am personally in favor of a little bit or verbiage (maybe with a link?) explaining what a sanitizer is.

@disconnect3d can you update this patch with a proposal?

**Comment by eholk** [comment]
Marking this as waiting on author so @disconnect3d can resolve conflicts and add a small note about what a sanitizer is.

@rustbot author

**Comment by Dylan-DPC** [comment]
@disconnect3d any updates on this? thanks

**Comment by disconnect3d** [comment]
@apiraino @eholk @Dylan-DPC Updated!

**Comment by apiraino** [comment]
thanks @disconnect3d . As mentioned in a [previous comment](https://github.com/rust-lang/rust/pull/108942#issuecomment-1462444505), I'll switch to waiting on a review to signal that this PR is ready for a review.

(For the record, merge commits are not allowed in PRs -- see [https://rustc-dev-guide.rust-lang.org/git.html#no-merge-policy](https://rustc-dev-guide.rust-lang.org/git.html#no-merge-policy))

@rustbot review

**Comment by Dylan-DPC** [comment]
@disconnect3d can you remove the merge commit and rebase instead? After that we will get it reviewed

**Comment by JohnCSimon** [comment]
@disconnect3d 
Ping from triage: I'm closing this due to inactivity, Please reopen when you are ready to continue with this. 
Note: if you are going to continue please open the PR **BEFORE** you push to it, else you won't be able to reopen - this is a quirk of github. 
Thanks for your contribution.

@rustbot label: +S-inactive

**Comment by disconnect3d** [comment]
@JohnCSimon just to be sure, how do I reopen this?

**Comment by JohnCSimon** [comment]
@disconnect3d 
image

I clicked "reopen pull request"

**Comment by oskgo** [comment]
It seems like these changes were merged in #121195, so I'll close this.
Thanks for your contribution.

**Pull request closed without merging**